File: go/internal/ntt/ntt.go
// Package ntt implements the Number Theoretic Transform for ML-KEM (FIPS 203).
//
// This includes the forward NTT (Algorithm 9), inverse NTT (Algorithm 10),
// pointwise NTT multiplication (Algorithm 11), and base-case multiply
// (Algorithm 12) as specified in FIPS 203.
package ntt

import (
	"github.com/baron-chain/PQC-Standards-Implementation/go/internal/field"
)

// Zetas contains the precomputed twiddle factors for the NTT.
// Zetas[i] = 17^BitRev7(i) mod 3329, computed at compile time.
var Zetas = [128]uint16{
	1, 1729, 2580, 3289, 2642, 630, 1897, 848,
	1062, 1919, 193, 797, 2786, 3260, 569, 1746,
	296, 2447, 1339, 1476, 3046, 56, 2240, 1333,
	1426, 2094, 535, 2882, 2393, 2879, 1974, 821,
	289, 331, 3253, 1756, 1197, 2304, 2277, 2055,
	650, 1977, 2513, 632, 2865, 33, 1320, 1915,
	2319, 1435, 807, 452, 1438, 2868, 1534, 2402,
	2647, 2617, 1481, 648, 2474, 3110, 1227, 910,
	17, 2761, 583, 2649, 1637, 723, 2288, 1100,
	1409, 2662, 3281, 233, 756, 2156, 3015, 3050,
	1703, 1651, 2789, 1789, 1847, 952, 1461, 2687,
	939, 2308, 2437, 2388, 733, 2337, 268, 641,
	1584, 2298, 2037, 3220, 375, 2549, 2090, 1645,
	1063, 319, 2773, 757, 2099, 561, 2466, 2594,
	2804, 1092, 403, 1026, 1143, 2150, 2775, 886,
	1722, 1212, 1874, 1029, 2110, 2935, 885, 2154,
}

// NTT performs the in-place forward Number Theoretic Transform on f.
// This is Algorithm 9 of FIPS 203.
func NTT(f *[256]field.Element) {
	k := 1
	for length := 128; length >= 2; length /= 2 {
		for start := 0; start < 256; start += 2 * length {
			zeta := field.New(Zetas[k])
			k++
			for j := start; j < start+length; j++ {
				t := zeta.Mul(f[j+length])
				f[j+length] = f[j].Sub(t)
				f[j] = f[j].Add(t)
			}
		}
	}
}

// NTTInverse performs the in-place inverse NTT on f.
// This is Algorithm 10 of FIPS 203.
func NTTInverse(f *[256]field.Element) {
	k := 127
	for length := 2; length <= 128; length *= 2 {
		for start := 0; start < 256; start += 2 * length {
			zeta := field.New(Zetas[k])
			k--
			for j := start; j < start+length; j++ {
				t := f[j]
				f[j] = t.Add(f[j+length])
				f[j+length] = zeta.Mul(f[j+length].Sub(t))
			}
		}
	}
	// Multiply every coefficient by 128^{-1} mod q = 3303.
	invN := field.New(3303)
	for i := 0; i < 256; i++ {
		f[i] = f[i].Mul(invN)
	}
}

// MultiplyNTTs computes the pointwise product of two NTT-domain polynomials.
// This is Algorithm 11 of FIPS 203.
func MultiplyNTTs(fHat, gHat *[256]field.Element) [256]field.Element {
	var h [256]field.Element
	for i := 0; i < 64; i++ {
		// First pair: indices (4i, 4i+1) with gamma = Zetas[64+i].
		gamma := field.New(Zetas[64+i])
		c0, c1 := baseCaseMultiply(
			fHat[4*i], fHat[4*i+1],
			gHat[4*i], gHat[4*i+1],
			gamma,
		)
		h[4*i] = c0
		h[4*i+1] = c1

		// Second pair: indices (4i+2, 4i+3) with gamma = -Zetas[64+i].
		c0, c1 = baseCaseMultiply(
			fHat[4*i+2], fHat[4*i+3],
			gHat[4*i+2], gHat[4*i+3],
			gamma.Neg(),
		)
		h[4*i+2] = c0
		h[4*i+3] = c1
	}
	return h
}

// baseCaseMultiply performs the degree-1 polynomial multiplication modulo
// (X^2 - gamma). This is Algorithm 12 of FIPS 203.
//
// Given a(X) = a0 + a1*X and b(X) = b0 + b1*X, it computes:
//
//	c0 = a0*b0 + a1*b1*gamma
//	c1 = a0*b1 + a1*b0
func baseCaseMultiply(a0, a1, b0, b1, gamma field.Element) (field.Element, field.Element) {
	c0 := a0.Mul(b0).Add(a1.Mul(b1).Mul(gamma))
	c1 := a0.Mul(b1).Add(a1.Mul(b0))
	return c0, c1
}

File: swift/Sources/PQCStandards/MlKem/Ntt.swift
// ML-KEM Number Theoretic Transform

public enum KemNtt {
    // Precomputed zetas: powers of 17 mod 3329 in bit-reversed order
    // zetas[i] = 17^(brv(i)) mod q, for i in 0..<128, with 7-bit bit-reversal
    public static let zetas: [Int32] = {
        let q: Int32 = 3329
        let g: Int32 = 17
        // First compute all powers of g mod q up to 256
        var powers = [Int32](repeating: 0, count: 256)
        powers[0] = 1
        for i in 1..<256 {
            powers[i] = Int32((Int64(powers[i-1]) * Int64(g)) % Int64(q))
        }
        func bitrev7(_ x: Int) -> Int {
            var r = 0; var v = x
            for _ in 0..<7 { r = (r << 1) | (v & 1); v >>= 1 }
            return r
        }
        var result = [Int32](repeating: 0, count: 128)
        for i in 0..<128 {
            result[i] = powers[bitrev7(i)]
        }
        return result
    }()

    // Gammas for basemul: gamma[i] = 17^(2*brv(i)+1) mod q for i in 0..<128
    // where brv is 7-bit reversal
    public static let gammas: [Int32] = {
        let q: Int32 = 3329
        let g: Int32 = 17
        var powers = [Int32](repeating: 0, count: 512)
        powers[0] = 1
        for i in 1..<512 {
            powers[i] = Int32((Int64(powers[i-1]) * Int64(g)) % Int64(q))
        }
        func bitrev7(_ x: Int) -> Int {
            var r = 0; var v = x
            for _ in 0..<7 { r = (r << 1) | (v & 1); v >>= 1 }
            return r
        }
        var result = [Int32](repeating: 0, count: 128)
        for i in 0..<128 {
            let exp = 2 * bitrev7(i) + 1
            result[i] = powers[exp % 256]
        }
        return result
    }()

    /// Forward NTT: in-place, 7 layers
    public static func ntt(_ f: inout [Int32]) {
        var k = 1
        var len = 128
        while len >= 2 {
            var start = 0
            while start < 256 {
                let z = zetas[k]
                k += 1
                for j in start..<(start + len) {
                    let t = KemField.mul(z, f[j + len])
                    f[j + len] = KemField.sub(f[j], t)
                    f[j] = KemField.add(f[j], t)
                }
                start += 2 * len
            }
            len /= 2
        }
    }

    /// Inverse NTT: in-place, 7 layers
    public static func invNtt(_ f: inout [Int32]) {
        var k = 127
        var len = 2
        while len <= 128 {
            var start = 0
            while start < 256 {
                let z = zetas[k]
                k -= 1
                for j in start..<(start + len) {
                    let t = f[j]
                    f[j] = KemField.add(t, f[j + len])
                    f[j + len] = KemField.mul(z, KemField.sub(f[j + len], t))
                }
                start += 2 * len
            }
            len *= 2
        }
        // Multiply by n^{-1} mod q: 256^{-1} mod 3329 = 3316
        // Since 3329 = 13*256 + 1, so 256 * (-13) = 1 (mod 3329), so 256^{-1} = 3329-13 = 3316
        let nInv: Int32 = 3316
        for i in 0..<256 {
            f[i] = KemField.mul(f[i], nInv)
        }
    }

    /// Basemul: multiply two NTT-domain polynomials pairwise
    public static func basemul(_ a: [Int32], _ b: [Int32]) -> [Int32] {
        var r = [Int32](repeating: 0, count: 256)
        for i in 0..<128 {
            let gamma = gammas[i]
            let a0 = a[2*i], a1 = a[2*i+1]
            let b0 = b[2*i], b1 = b[2*i+1]
            // (a0 + a1*X) * (b0 + b1*X) mod (X^2 - gamma)
            r[2*i]   = KemField.add(KemField.mul(a0, b0), KemField.mul(KemField.mul(a1, b1), gamma))
            r[2*i+1] = KemField.add(KemField.mul(a0, b1), KemField.mul(a1, b0))
        }
        return r
    }

    /// Add two polynomials
    public static func polyAdd(_ a: [Int32], _ b: [Int32]) -> [Int32] {
        var r = [Int32](repeating: 0, count: 256)
        for i in 0..<256 {
            r[i] = KemField.add(a[i], b[i])
        }
        return r
    }

    /// Subtract two polynomials
    public static func polySub(_ a: [Int32], _ b: [Int32]) -> [Int32] {
        var r = [Int32](repeating: 0, count: 256)
        for i in 0..<256 {
            r[i] = KemField.sub(a[i], b[i])
        }
        return r
    }
}
